hail/src/main/scala/is/hail/expr/ir/lowering/LowerTableIR.scala

```diff
@@ -19,86 +19,121 @@ object TableStage {
-
-  def partition(ctxRef: Ref): IR
-
-  def wrapInBindings(body: IR): IR = {
-    letBindings.foldRight(body) { case ((name, binding), soFar) => Let(name, binding, soFar) }
+  val contexts: IR,
+  val ctxRefName: String,
+  val partitionIR: IR) { self =>
+
+  def ctxType: Type = contexts.typ.asInstanceOf[TStream].elementType
+  def rowType: TStruct = partitionIR.typ.asInstanceOf[TStream].elementType.asInstanceOf[TStruct]
+  def kType: TStruct = partitioner.kType
+  def key: IndexedSeq[String] = kType.fieldNames
+  def globalType: TStruct = globals.typ.asInstanceOf[TStruct]
+
+  assert(kType.fields.forall(f => rowType.field(f.name).typ == f.typ))
+  assert(broadcastVals.exists { case (name, value) => name == globals.name && value == globals})
+
+  def copy(
+    letBindings: IndexedSeq[(String, IR)] = letBindings,
+    broadcastVals: IndexedSeq[(String, IR)] = broadcastVals,
+    globals: Ref = globals,
+    partitioner: RVDPartitioner = partitioner,
+    contexts: IR = contexts,
+    ctxRefName: String = ctxRefName,
+    partitionIR: IR = partitionIR
+  ): TableStage =
+    new TableStage(letBindings, broadcastVals, globals, partitioner, contexts, ctxRefName, partitionIR)
+
+  def partition(ctx: IR): IR = {
+    require(ctx.typ == ctxType)
+    Let(ctxRefName, ctx, partitionIR)
   }
 
-  def mapPartition(f: IR => IR): TableStage = {
-    val outer = this
-    new TableStage(letBindings, broadcastVals, globals, partitioner, contexts) {
-      def partition(ctxRef: Ref): IR = f(outer.partition(ctxRef))
-    }
+  def wrapInBindings(body: IR): IR = letBindings.foldRight[IR](body) {
+    case ((name, value), body) => Let(name, value, body)
   }
 
+  def mapPartition(f: IR => IR): TableStage =
+    copy(partitionIR = f(partitionIR))
+
   def mapContexts(f: IR => IR): TableStage = {
-    val outer = this
-    new TableStage(letBindings, broadcastVals, globals, partitioner, f(contexts)) {
-      override def partition(ctxRef: Ref): IR = outer.partition(ctxRef)
-    }
+    val newContexts = f(contexts)
+    assert(newContexts.typ == contexts.typ)
+
+    copy(contexts = newContexts)
   }
 
   def mapGlobals(f: IR => IR): TableStage = {
     val newGlobals = f(globals)
-    val newID = genUID()
-    val outer = this
-    new TableStage(
-      letBindings :+ newID -> newGlobals,
-      broadcastVals + newID,
-      Ref(newID, newGlobals.typ),
-      partitioner, contexts) {
-      def partition(ctxRef: Ref): IR = outer.partition(ctxRef)
-    }
+    val globalsRef = Ref(genUID(), newGlobals.typ)
+
+    copy(
+      letBindings = letBindings :+ globalsRef.name -> newGlobals,
+      broadcastVals = broadcastVals :+ globalsRef.name -> globalsRef,
+      globals = globalsRef)
   }
 
-  def collect(bind: Boolean = true): IR = {
-    val ctx = Ref(genUID(), types.coerce[TStream](contexts.typ).elementType)
-    val broadcastRefs = MakeStruct(letBindings.filter { case (name, _) => broadcastVals.contains(name) })
+  def mapCollect(f: IR => IR): IR = {
```

Comment:
I guess I figured if any user is forced to call `wrapInBindings`, why not just be consistent and always have to. One of the things that I first found confusing about `TableStage` was that it wasn't obvious what added bindings and what didn't. It seemed cleaner in some ways to just always make it explicit. (I really wanted to get rid of `letBindings` and `wrapInBindings` entirely, but I don't think that's possible with local sort as it is currently.)

I refactored it so that `wrapInBindings` can be private, by adding `getGlobals` and `getNumPartitions`, that return self-contained IRs. Still not completely satisfactory, but I think fine for now.